src/engine/Util/WebAudio.ts

```diff
@@ -8,41 +9,53 @@ export class WebAudio {
    * right after a user interaction event. Typically used by [[PauseAfterLoader]]
    * @source https://paulbakaus.com/tutorials/html5/web-audio-on-ios/
    */
-  static unlock() {
+  static unlock(): Promise<boolean> {
+    let promise = new Promise<boolean>();
     if (WebAudio._unlocked || !AudioContextFactory.create()) {
-      return;
+      return promise.resolve(true);
     }
 
     const audioContext = AudioContextFactory.create();
-    // create empty buffer and play it
-    const buffer = audioContext.createBuffer(1, 1, 22050);
-    const source = audioContext.createBufferSource();
-    let ended = false;
-
-    source.buffer = buffer;
-    source.connect(audioContext.destination);
-    source.onended = () => (ended = true);
-
-    if ((<any>source).noteOn) {
-      // deprecated
-      (<any>source).noteOn(0);
-    } else {
-      source.start(0);
-    }
+    audioContext.resume().then(
+      () => {
+        // create empty buffer and play it
+        const buffer = audioContext.createBuffer(1, 1, 22050);
+        const source = audioContext.createBufferSource();
+        let ended = false;
 
-    // by checking the play state after some time, we know if we're really unlocked
-    setTimeout(function() {
-      if ((<any>source).playbackState) {
-        var legacySource = <any>source;
-        if (legacySource.playbackState === legacySource.PLAYING_STATE || legacySource.playbackState === legacySource.FINISHED_STATE) {
-          WebAudio._unlocked = true;
-        }
-      } else {
-        if (audioContext.currentTime > 0 || ended) {
-          WebAudio._unlocked = true;
+        source.buffer = buffer;
+        source.connect(audioContext.destination);
+        source.onended = () => (ended = true);
+
+        if ((<any>source).noteOn) {
+          // deprecated
+          (<any>source).noteOn(0);
+        } else {
+          source.start(0);
         }
+
+        // by checking the play state after some time, we know if we're really unlocked
+        setTimeout(() => {
+          if ((<any>source).playbackState) {
```

Comment:
So, I am guessing the casting to `any` is due to a legacy feature where these properties may exist there but don't anymore. I would suggest instead of this checking, we leverage [user-defined type guarding](https://basarat.gitbooks.io/typescript/content/docs/types/typeGuard.html) to make a `LegacyWebAudioSource` interface:

```ts
function isLegacyWebAudioSource(source: any): source is LegacyWebAudioSource {
  return source.playbackState;
}
```

That way, you can strongly type the legacy properties under that `if` condition (or elsewhere!)